lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -1635,15 +1636,64 @@ balanceTransaction ctx change pp ti wallet unadjustedPtx = do
                 @era @m @s @k @ktype
                 ctx change pp ti wallet strategy adjustedPtx
     balanceWith SelectionStrategyOptimal
-        `catchE` \case
-            ErrBalanceTxMaxSizeLimitExceeded
-                -> balanceWith SelectionStrategyMinimal
-            otherErr
-                -> throwE otherErr
+        `catchE` \e ->
+            if minimalStrategyIsWorthTrying e
+            then balanceWith SelectionStrategyMinimal
+            else throwE e
   where
     shelleyEra = Cardano.shelleyBasedEra @era
     recentEra = WriteTx.recentEra @era
 
+    -- Determines whether or not the minimal selection strategy is worth trying.
+    -- This depends upon the way in which the optimal selection strategy failed.
+    minimalStrategyIsWorthTrying :: ErrBalanceTx -> Bool
+    minimalStrategyIsWorthTrying e = or
+        [ maxSizeLimitExceeded
+        , unableToConstructChange
+        , selectionCollateralError
+        ]
+      where
+        -- The size of a transaction can be reduced by selecting fewer inputs,
+        -- or by generating less change. Since the minimal selection strategy
+        -- selects as few inputs and generates as little change as possible,
+        -- using this strategy might allow us to generate a transaction within
+        -- the size limit.
+        maxSizeLimitExceeded = case e of
+            ErrBalanceTxMaxSizeLimitExceeded ->
+                True
+            _someOtherError ->
+                False
+
+        -- In situations where the available supply of ada is constrained, or
+        -- where all available ada is bundled up with other tokens, this can
+        -- prevent us from generating change. In this case, trying again with
+        -- the minimal selection strategy might allow us to select fewer
+        -- inputs, generate less change, lower the amount of ada required to
+        -- pay for the change, and therefore increase the chance that we can
+        -- generate change successfully.
+        unableToConstructChange = case e of
+            ErrBalanceTxSelectAssets
+                (ErrSelectAssetsSelectionError
+                (SelectionBalanceErrorOf
+                (UnableToConstructChange {}))) ->
```

Comment:
👌 Though I don't quite understand `UnableToConstructChange`, particularly in relation to `BalanceInsufficient`. (They look [similar](https://github.com/input-output-hk/cardano-wallet/blob/081fc68fa9591ccb124f38c720d0ddecc73024f9/lib/wallet/test/data/balanceTx/1ada-payment/golden#L20-L25) to me) Are you sure this may have an effect?

Reply:
It would be interesting to see a unit/golden test showing the fallback helping for this error at some point. For txMaxSize I wrote these: https://github.com/input-output-hk/cardano-wallet/blob/e52d1b378e00776b9c3b6ffa3553dc1cae1b2c6e/lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs#L2389-L2408

Not saying we need this now, but it would be interesting.

Reply:
> Not saying we need this now, but it would be nice/interesting.

I'll add this to the ticket (ADP-2488).

Reply:
> Though I don't quite understand UnableToConstructChange, particularly in relation to BalanceInsufficient. (They look somewhat [similar](https://github.com/input-output-hk/cardano-wallet/blob/081fc68fa9591ccb124f38c720d0ddecc73024f9/lib/wallet/test/data/balanceTx/1ada-payment/golden#L20-L25) to me, though I understand they are related to different stages of the coin-selection process) Are you sure this may have an effect?

I'll try to answer this tomorrow. I've added a reminder to the ticket to come back and look at this!